File: 0113.PathSumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        ans = []
        self.dfs(root, sum, [], ans)
        return ans
        
    def dfs(self, root, sum, path, ans):
        if not root:
            return
        if not root.left and not root.right:
            if sum == root.val:
                ans.append(path+[root.val])
            return
        self.dfs(root.left, sum-root.val, path+[root.val], ans)
        self.dfs(root.right, sum-root.val, path+[root.val], ans)
        

File: 0113.PathSumII.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func pathSum(root *TreeNode, sum int) [][]int {
    if root == nil {
        return nil
    }
    ans := [][]int{}
    dfs(root, sum, []int{}, &ans)
    return ans
}

func dfs(node *TreeNode, sum int, path []int, ans *[][]int) {
    if node.Left == nil && node.Right == nil && sum == node.Val {
        tmp := make([]int, len(path)+1)
        copy(tmp, path)
        tmp[len(path)] = node.Val
        *ans = append(*ans, tmp)
    }
    if node.Left != nil {
        dfs(node.Left, sum-node.Val, append(path, node.Val), ans)
    }
    if node.Right != nil {
        dfs(node.Right, sum-node.Val, append(path, node.Val), ans)
    }
}